core/templates/variant_allocator.h

```diff
@@ -0,0 +1,294 @@
+			IN_AVAILABLE,
+		};
+
+		Item objects[OBJECTS_PER_SLAB];
+		uint64_t bitmap = UINT64_MAX;
+		SpinLock lock;
+		Slab *next = nullptr;
+		Slab *next_available = nullptr;
+		std::atomic<slabstate> state{ IN_USE };
+
+		template <typename... Args>
+		T *allocate(Args &&...p_args) {
+			lock.lock();
+			// Under lock, nobody else is writing. Locking/unlocking is a barrier.
+			if (bitmap != 0) {
+				uint8_t index = find_first_trailing_set_bit64(bitmap);
+				bitmap &= ~(1ULL << index);
+				lock.unlock();
+				objects[index].index = index;
+				if (!std::is_trivially_constructible<T>()) {
+					new (&(objects[index].object)) T(std::forward<Args>(p_args)...);
+				}
+				return (T *)&(objects[index].object);
+			}
+			// We're about to replace our slab.
+			state = UNUSED;
+			lock.unlock();
+			return nullptr;
+		}
+
+		void deallocate(Item *ptr) {
+			if (!std::is_trivially_destructible<T>()) {
+				(*((T *)&ptr->object)).~T();
+			}
+			lock.lock();
+			bitmap |= (1ULL << ptr->index);
+			uint8_t usage = count_set_bits64(bitmap);
+
+			// state cannot change while we're under lock.
+			if (usage <= 16 && state == UNUSED) {
+				avail_spin_lock.lock();
+				// Under lock, nobody else is writing. Locking/unlocking is a barrier.
+				Slab *global = global_usable_slabs;
+				next_available = global;
+				state = IN_AVAILABLE;
+				global_usable_slabs = this;
+				avail_spin_lock.unlock();
+			}
+
+			lock.unlock();
+		}
+
+		void claim() {
+			// Not under lock, use atomcis.
+			state.store(IN_USE, std::memory_order_release);
+		}
+	};
+
+	static thread_local Slab *local_slab;
+	static Slab *global_slabs;
+	static Slab *global_usable_slabs;
+	static SpinLock alloc_spin_lock;
+	static SpinLock avail_spin_lock;
+	static std::atomic<size_t> allocator_count;
+
+	Slab *allocate_slab() {
+		// Not under lock, guessing.
+		if (global_usable_slabs) {
+			avail_spin_lock.lock();
+			// Under lock, nobody else is writing. Locking/unlocking is a barrier.
+			if (global_usable_slabs) {
+				Slab *new_slab = global_usable_slabs;
+				Slab *next_global_usable = global_usable_slabs->next_available;
+				global_usable_slabs = next_global_usable;
+				avail_spin_lock.unlock();
+
+				new_slab->claim();
+				return new_slab;
+			}
+			avail_spin_lock.unlock();
+		}
+
+		void *raw = std::malloc(sizeof(Slab));
+		Slab *slab = new (raw) Slab();
+
+		alloc_spin_lock.lock();
+		slab->next = global_slabs;
+		global_slabs = slab;
+		alloc_spin_lock.unlock();
+
+		return slab;
+	}
+
+#if defined(builtin_popcountll)
+	_ALWAYS_INLINE_ static int popcount64(uint64_t x) {
+		return __builtin_popcountll(x);
+	}
+#else
+	// Hamming weight calculation version of popcount.
+	_ALWAYS_INLINE_ static int popcount64(uint64_t x) {
+		x = x - ((x >> 1) & 0x5555555555555555);
+		x = (x & 0x3333333333333333) + ((x >> 2) & 0x3333333333333333);
+		x = (x + (x >> 4)) & 0x0F0F0F0F0F0F0F0F;
+		x = x + (x >> 8);
+		x = x + (x >> 16);
+		x = x + (x >> 32);
+		return x & 0x7F;
+	}
+#endif
```

Comment:
```suggestion
	// Hamming weight calculation version of popcount.
	_ALWAYS_INLINE_ static int popcount64(uint64_t p_word) {
#ifdef builtin_popcountg
		return __builtin_popcountg(p_word);
#else
		p_word = p_word - ((p_word >> 1) & 0x5555555555555555);
		p_word = (p_word & 0x3333333333333333) + ((p_word >> 2) & 0x3333333333333333);
		p_word = (p_word + (p_word >> 4)) & 0x0F0F0F0F0F0F0F0F;
		p_word = p_word + (p_word >> 8);
		p_word = p_word + (p_word >> 16);
		p_word = p_word + (p_word >> 32);
		return p_word & 0x7F;
#endif // builtin_popcountg
	}
```
Some of the operations look like they can be further simplified — `p_word = p_word - ...` → `p_word -= ...` — but the syntax could be deliberate or I'm maybe missing something, so I left that alone